### ISSUE TITLE
SameDiff: creating gradient function can break forward pass?

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#5400]
```
    @Test
    public void gradFnBug() {
        Nd4j.setDataType(DataBuffer.Type.DOUBLE);
        SameDiff sd = SameDiff.create();

        int nOut = 4;
        int minibatch = 10;
        SDVariable input = sd.var("in", new int[]{-1, nOut});

        SDVariable loss = sd.mean("loss", input);

        INDArray inputArr = Nd4j.rand(minibatch, nOut);
        sd.associateArrayWithVariable(inputArr, input);

        sd.createGradFunction();
        sd.execAndEndResult();  //Exception
    }
```

```
org.nd4j.linalg.exception.ND4JIllegalStateException: Array for loss does not exist. Please use putArrayForVertexId instead.

	at org.nd4j.autodiff.samediff.SameDiff.updateArrayForVarName(SameDiff.java:545)
	at org.nd4j.autodiff.samediff.SameDiff.exec(SameDiff.java:6133)
	at org.nd4j.autodiff.samediff.SameDiff.execAndEndResult(SameDiff.java:4926)
	at org.nd4j.autodiff.gradcheck.TempDebug.testReductionGradientsSimple2(TempDebug.java:81)
```

Executes fine without the createGradFunction() call first...
